data/transactions/logic/eval.go

```diff
@@ -4060,3 +4062,126 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+func hasDuplicateKeys(dec *json.Decoder) (bool, error) {
+	keys := make(map[string]bool)
+	var value interface{}
+	dec.Token() // {
+	for dec.More() {
+		// Get JSON key
+		key, err := dec.Token()
+		if err != nil {
+			return false, err
+		}
+		// end of json
+		if key == '}' {
+			break
+		}
+		// decode value
+		err = dec.Decode(&value)
+		if err != nil {
+			return false, err
+		}
+		// check for duplicates
+		if _, ok := keys[key.(string)]; ok {
+			return true, nil
+		}
+		keys[key.(string)] = true
+	}
+	return false, nil
+}
+
+func parseJSON(jsonText []byte) (interface{}, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text")
+	}
+	// check for duplicate keys
+	decoder := json.NewDecoder(bytes.NewReader(jsonText))
+	hasDuplicates, err := hasDuplicateKeys(decoder)
+	if hasDuplicates {
+		return nil, fmt.Errorf("invalid json text, duplicate keys not allowed")
+	}
+	// decode text
+	decoder = json.NewDecoder(bytes.NewReader(jsonText))
+	decoder.UseNumber()
+	var parsed interface{}
+	err = decoder.Decode(&parsed)
+	if err != nil {
+		return nil, err
+	}
+	return parsed, nil
+}
+func opJSONRef(cx *EvalContext) {
+	// get json key
+	last := len(cx.stack) - 1
+	key := string(cx.stack[last].Bytes)
+	cx.stack = cx.stack[:last] // pop
+
+	//parse json text
+	last = len(cx.stack) - 1
+	resp, err := parseJSON(cx.stack[last].Bytes)
+	if err != nil {
+		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
+		return
+	}
+
+	//get value from json
+	var stval stackValue
+	resp, ok := resp.(map[string]interface{})[key]
+	if !ok {
+		cx.err = fmt.Errorf("key %s not found in JSON text", key)
+		return
+	}
+	expectedType := jsonRefTypeNames[cx.program[cx.pc+1]]
+	switch expectedType {
+	case "JSONString":
```

Comment:
nit: compare `cx.program[cx.pc+1]` directly to the constants `JSONString`, `JSONUint64`, and `JSONObject` instead